sdk/search/search-documents/src/indexModels.ts

```diff
@@ -685,3 +722,136 @@ export type IndexDocumentsAction<T> = {
 } & Partial<T>;
 
 // END manually modified generated interfaces
+
+// Utility types
+
+export type UnionToIntersection<U> =
+  // Distribute members of U into parameter position of a union of functions
+  (
+    U extends unknown ? (_: U) => unknown : never
+  ) extends // Infer the intersection of the members of U as a single intersected parameter type
+  (_: infer I) => unknown
+    ? I
+    : never;
+
+// Types that should not be included in SelectFields recursion
+export type ExcludedODataTypes = Date | GeographyPoint;
+
+/**
+ * Produces a union of valid Cognitive Search OData $select paths for T
+ * using a post-order traversal of the field tree rooted at T.
+ */
+export type SelectFields<T extends object> = T extends Array<infer U>
+  ? // Allow selecting fields only from elements which are objects
+    NonNullable<U> extends object
+    ? SelectFields<NonNullable<U>>
+    : never
+  : {
+      // Only consider string keys
+      [K in Exclude<keyof T, symbol | number>]: NonNullable<T[K]> extends object
+        ? NonNullable<T[K]> extends ExcludedODataTypes
+          ? // Excluded, so don't recur
+            K
+          : SelectFields<NonNullable<T[K]>> extends infer NextPaths
+          ? // Narrow NextPaths' type for template literal
+            NextPaths extends string
+            ? // Union this key with all the next paths separated with '/'
+              K | `${K}/${NextPaths}`
+            : // We didn't infer any nested paths, so just use this key
+              K
+          : never
+        : // Not an object, so can't recur
+          K;
+    }[Exclude<keyof T, symbol | number>];
+
+/**
+ * Deeply pick fields of T using valid Cognitive Search OData $select
+ * paths.
+ */
+export type SearchPick<T extends object, Paths extends SelectFields<T>> = [T] extends [never]
```

Comment:
Is there any valid scenario where the IndexModel has a property of type IndexModel? Like a parent?

If it is a valid scenario this type would break

```ts
interface Foo {
      foo: string,
      parent?: Foo
}
```